nativelink-service/src/bytestream_server.rs

```diff
@@ -227,23 +237,51 @@ impl ByteStreamServer {
         Server::new(self).max_decoding_message_size(max_decoding_message_size)
     }
 
+    /// Creates or joins an upload stream for the given UUID.
+    ///
+    /// This function handles three scenarios:
+    /// 1. UUID doesn't exist - creates a new upload stream
+    /// 2. UUID exists but is idle - resumes the existing stream
+    /// 3. UUID exists and is active - generates a unique UUID by appending a nanosecond
+    ///    timestamp to avoid collision, then creates a new stream with that UUID
+    ///
+    /// The nanosecond timestamp ensures virtually zero probability of collision since
+    /// two concurrent uploads would need to both collide on the original UUID AND
+    /// generate the unique UUID in the exact same nanosecond.
     fn create_or_join_upload_stream(
         &self,
-        uuid: String,
+        uuid: &str,
         store: Store,
         digest: DigestInfo,
-    ) -> Result<ActiveStreamGuard<'_>, Error> {
-        let (uuid, bytes_received) = match self.active_uploads.lock().entry(uuid) {
+    ) -> ActiveStreamGuard<'_> {
+        let (uuid, bytes_received) = match self.active_uploads.lock().entry(uuid.to_string()) {
             Entry::Occupied(mut entry) => {
                 let maybe_idle_stream = entry.get_mut();
-                let Some(idle_stream) = maybe_idle_stream.1.take() else {
-                    return Err(make_input_err!("Cannot upload same UUID simultaneously"));
-                };
-                let bytes_received = maybe_idle_stream.0.clone();
-                info!(msg = "Joining existing stream", entry = ?entry.key());
-                return Ok(idle_stream.into_active_stream(bytes_received, self));
+                if let Some(idle_stream) = maybe_idle_stream.1.take() {
+                    // Case 2: Stream exists but is idle, we can resume it
+                    let bytes_received = maybe_idle_stream.0.clone();
+                    info!(msg = "Joining existing stream", entry = ?entry.key());
+                    return idle_stream.into_active_stream(bytes_received, self);
+                }
+                // Case 3: Stream is active - generate a unique UUID to avoid collision
+                // Using nanosecond timestamp makes collision probability essentially zero
+                let original_uuid = entry.key().clone();
+                let unique_uuid = Self::generate_unique_uuid(&original_uuid);
```

Comment:
Wondering about this one. Can't we just make an entirely new UUID rather than doing the timestamp thing? I can see some benefits of the timestamp stuff (mostly to go "these two streams should have been the same thing" from eyeballing the logs), but a new fresh UUID is definitely unique and we don't have to even worry about the unlikely timestamp case.

Reply:
We're doing this so we can more easily track the real problem later on v.s. hiding it entirely with new uuids.